std/range/interfaces.d

```diff
@@ -377,7 +383,10 @@ if (isInputRange!(Unqual!R))
             @property E front() { return _range.front; }
 
             E moveFront() {
-                return _range.moveFront();
+                static if (__traits(compiles, _range.moveFront()))
+                    return _range.moveFront();
+                else
+                    throw new Exception("Cannot move the front of a(n) `" ~ R.stringof ~ "`");
```

Comment:
As discussed in previous comments, the interface method that this implements is not `@nogc` or `nothrow`, so making this `@nogc` or `nothrow` would have little benefit (although it could be easily be made `@nogc` by statically allocating the exception).

I used a generic `Exception` because it was simpler than introducing a new exception type. If we think users will want to distinguish between this exception and other exceptions that may be thrown by the call to `_range.moveFront()` (e.g., perhaps they can fall back to regular `front` if `moveFront` is not supported), then we should use a more specific exception type here.

Reply:
Exception do have the disadvantage that they are not as performant as if we did a separate `bool hasMobileElements()` or something like that. And they are not -betterC/LWDR friendly, although I suspect this module has quite a way to go for that anyway. But they are the simplest solution anyway. I'm fairly even on this.

Reply:
Unfortunately, adding a new virtual method to an interface is a breaking change, so that is not really an option here.

Reply:
> As discussed in previous comments, the interface method that this implements is not `@nogc` or `nothrow`, so making this `@nogc` or `nothrow` would have little benefit (although it could be easily be made `@nogc` by statically allocating the exception).

I just raised that because exceptions are exceptionally slow and should be avoided, although I kinda see the point here.

> I used a generic `Exception` because it was simpler than introducing a new exception type. If we think users will want to distinguish between this exception and other exceptions that may be thrown by the call to `_range.moveFront()` (e.g., perhaps they can fall back to regular `front` if `moveFront` is not supported), then we should use a more specific exception type here.

The exceptions are intended to be caught by the user and catching a generic one in normal code is wrong, unless you are doing some specific logging or other code that specifically requires it. See https://cwe.mitre.org/data/definitions/396.html

Reply:
> I just raised that because exceptions are exceptionally slow and should be avoided, although I kinda see the point here.

If you know of a way to avoid them that would actually work here, please tell us.

> The exceptions are intended to be caught by the user and catching a generic one in normal code is wrong, unless you are doing some specific logging or other code that specifically requires it.

Anyone calling `moveFront` is going to have to resort to `catch (Exception)` anyway, since there is no restriction on what kind of exceptions a range's `front` method is allowed to throw. The only reason to use a more specific exception here is to allow the user to distinguish between exceptions thrown by `InputRangeObject` and exceptions thrown by the wrapped range.

Which is probably a good enough reason, actually—I'll go ahead and make the change.